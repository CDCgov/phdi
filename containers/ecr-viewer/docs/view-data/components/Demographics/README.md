[**ecr-viewer**](../../../README.md) â€¢ **Docs**

***

[ecr-viewer](../../../README.md) / view-data/components/Demographics

# view-data/components/Demographics

## Index

### Functions

- [default](functions/default.md)
