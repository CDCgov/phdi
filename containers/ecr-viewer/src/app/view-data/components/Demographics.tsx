import { evaluate } from "fhirpath";
import { Bundle } from "fhir/r4";
import { PathMappings } from "@/app/utils";

interface DemographicsProps {
    fhirPathMappings: PathMappings
    fhirBundle: Bundle | undefined
}



const patientName = (fhirBundle: Bundle | undefined, fhirPathMappings: PathMappings) => {
    const givenNames = evaluate(fhirBundle, fhirPathMappings.patientGivenName).join(" ");
    const familyName = evaluate(fhirBundle, fhirPathMappings.patientFamilyName);

    return `${givenNames} ${familyName}`;
}


const Demographics = (
    { fhirPathMappings, fhirBundle }: DemographicsProps
) => {
    return (
        <div>
            <div
                className="padding-bottom-3"
                aria-labelledby="summary-box-key-information"
            >
                <div className="usa-summary-box__body">
                    <h3
                        className="usa-summary-box__heading padding-y-105"
                        id="summary-box-key-information"
                    >
                        Demographics
                    </h3>
                    <div className="usa-summary-box__text">
                        <div className="grid-row">
                            <div className="data-title"><h4>Patient Name</h4></div>
                            <div className="grid-col-auto">
                                {patientName(fhirBundle, fhirPathMappings)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Patient ID</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientId)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>DOB</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientDOB)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Sex</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientGender)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Race</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientRace)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Ethnicity</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientEthnicity)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Preferred Language</h4></div>
                            <div className="grid-col-auto">
                                {evaluate(fhirBundle, fhirPathMappings.patientLanguage)}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Patient Address</h4></div>
                            <div className="grid-col-auto text-pre-line">
                                {evaluate(fhirBundle, fhirPathMappings.patientStreetAddress).join("\n")} {'\n'}
                                {evaluate(fhirBundle, fhirPathMappings.patientCity)}, {evaluate(fhirBundle, fhirPathMappings.patientState)} {'\n'}
                                {evaluate(fhirBundle, fhirPathMappings.patientZipCode)}, USA
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                        <div className="grid-row">
                            <div className="data-title"><h4>Contact</h4></div>
                            <div className="grid-col-auto text-pre-line">
                                {evaluate(fhirBundle, fhirPathMappings.patientPhoneNumbers).map(phoneNumber => {
                                    return `tel: (${phoneNumber.use}) ${phoneNumber.value}`
                                }).join("\n")}
                                {'\n'}
                                {evaluate(fhirBundle, fhirPathMappings.patientEmails).map(email => {
                                    return `email: ${email.value}`
                                }).join("\n")}
                            </div>
                        </div>
                        <div className={"section__line_gray"} />
                    </div>
                </div>
            </div>
        </div>);
};

export default Demographics;