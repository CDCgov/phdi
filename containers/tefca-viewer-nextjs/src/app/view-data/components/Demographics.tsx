import { Patient, HumanName, Address, ContactPoint, Identifier } from 'fhir/r4';
import { DisplayData } from '@/app/utils';
import { DataDisplay } from '@/app/utils';
import * as dateFns from "date-fns";
import {evaluate} from "fhirpath";

export interface DemographicsProps {
    patient: Patient;
}

function formatDemographics(patient: Patient) {

    const demographicData: DisplayData[] = [
        {
          title: "Patient Name",
          value: formatName(patient.name ?? []),
        },
        { 
          title: "DOB",
          value: patient.birthDate ?? "",
        },
        {
          title: "Current Age",
          value: calculatePatientAge(patient)?.toString(),
        },
        { 
          title: "Sex",
          value: patient.gender ?? "",
        },
        { 
          title: "Race",
          value: evaluate(patient, "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.first().valueCoding.display")[0],
        },
        {
          title: "Ethnicity",
          value: evaluate(patient, "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.first().valueCoding.display")[0],
        },
        {
          title: "Tribal Affiliation",
          value: evaluate(patient, "Patient.extension.where(url='http: //hl7.org/fhir/us/ecr/StructureDefinition/us-ph-tribal-affiliation-extension').extension.where(url='TribeName').value.display")[0],
        },
        {
          title: "Preferred Language",
          value: evaluate(patient, "Patient.communication.first().language.coding.first().display")[0],
        },
        {
          title: "Address",
          value: formatAddress(patient.address ?? []),
        },
        {
          title: "Contact",
          value: formatContact(patient.telecom ?? []),
        },
        {
          title: "Patient Identifiers",
          value: formatIdentifier(patient.identifier ?? []),
        },
      ];

    return demographicData;
}

function formatName(names: HumanName[]) {
    let name = "";
    if (names.length > 0) {
        name = names[0].given?.join(" ") + " " + names[0].family;
    }
    return name;
}

function formatAddress(address: Address[]) {
    let formattedAddress = "";
    if (address.length > 0) {
        formattedAddress = address[0]?.line?.join("\n") + "\n" + address[0].city + ", " + address[0].state + " " + address[0].postalCode;
    }
    return formattedAddress;
}

function formatContact(contacts: ContactPoint[]) {
    return (
        contacts.map((contact) => {
            if (contact.system === "phone") {
                return `${contact.use}: ${contact.value}`;
            }
            else if (contact.system === "email") {
                return contact.value;
            }
        }
        ).join("\n")
    )
}

function formatIdentifier(identifier: Identifier[]) {
    return (
        identifier.map((id) => {
            let idType = id.type?.coding?.[0].display ?? "";
            if (idType === "") {
                idType = id.type?.text ?? "";
            }

            return `${idType}: ${id.value}`;
        }
        ).join("\n")
    )
}

export const calculatePatientAge = (patient: Patient) => {

    if (patient.birthDate) {
      const patientDOB = new Date(patient.birthDate);
      const today = new Date();
      return dateFns.differenceInYears(today, patientDOB);
    }
  };

export default function Demographics(props: DemographicsProps) {

    const demographicData = formatDemographics(props.patient);

    return (
        <div>
            {demographicData.map((item, index) => (
                <DataDisplay item={item} key={index} />
            ))}
        </div>
    );

}