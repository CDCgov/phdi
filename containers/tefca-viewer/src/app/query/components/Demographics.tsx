import { Patient, HumanName, Address, ContactPoint, Identifier } from "fhir/r4";
import { DisplayData } from "@/app/utils";
import { DataDisplay } from "@/app/utils";
import * as dateFns from "date-fns";
import { evaluate } from "fhirpath";

/**
 * Displays the demographic information of a patient.
 * @param {Patient} patient - The patient to display demographic information for.
 * @returns {React.FC} The Demographics component.
 */
export interface DemographicsProps {
  patient: Patient;
}

/**
 * Displays the demographic information of a patient.
 * @param props - The props for the Demographics component.
 * @param props.patient - The patient resource to display demographic information for.
 * @returns The Demographics component.
 */
const Demographics: React.FC<DemographicsProps> = ({ patient }) => {
  const demographicData = formatDemographics(patient);

  return (
    <div>
      {demographicData.map((item, index) => (
        <DataDisplay item={item} key={index} />
      ))}
    </div>
  );
};

export default Demographics;

/**
 * Formats the demographic information of a patient.
 * @param patient - The patient to format demographic information for.
 * @returns The formatted demographic information as an array of DisplayData objects.
 */
function formatDemographics(patient: Patient): DisplayData[] {
  const demographicData: DisplayData[] = [
    {
      title: "Patient Name",
      value: formatName(patient.name ?? []),
    },
    {
      title: "DOB",
      value: patient.birthDate ?? "",
    },
    {
      title: "Current Age",
      value: calculatePatientAge(patient)?.toString(),
    },
    {
      title: "Sex",
      value: patient.gender ?? "",
    },
    {
      title: "Race",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Ethnicity",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Tribal Affiliation",
      value: evaluate(
        patient,
        "Patient.extension.where(url='http: //hl7.org/fhir/us/ecr/StructureDefinition/us-ph-tribal-affiliation-extension').extension.where(url='TribeName').value.display",
      )[0],
    },
    {
      title: "Preferred Language",
      value: evaluate(
        patient,
        "Patient.communication.first().language.coding.first().display",
      )[0],
    },
    {
      title: "Address",
      value: formatAddress(patient.address ?? []),
    },
    {
      title: "Contact",
      value: formatContact(patient.telecom ?? []),
    },
    {
      title: "Patient Identifiers",
      value: formatIdentifier(patient.identifier ?? []),
    },
  ];

  return demographicData;
}

/**
 * Formats the name of a FHIR HumanName object.
 * @param names - The HumanName object to format.
 * @returns The formatted name.
 */
function formatName(names: HumanName[]): string {
  let name = "";
  if (names.length > 0) {
    name = names[0].given?.join(" ") + " " + names[0].family;
  }
  return name;
}

/**
 * Formats the address of a FHIR Address object.
 * @param address - The Address object to format.
 * @returns The formatted address.
 */
function formatAddress(address: Address[]): string {
  let formattedAddress = "";
  if (address.length > 0) {
    formattedAddress =
      address[0]?.line?.join("\n") +
      "\n" +
      address[0].city +
      ", " +
      address[0].state +
      " " +
      address[0].postalCode;
  }
  return formattedAddress;
}

/**
 * Formats the contact information of a FHIR ContactPoint object.
 * @param contacts - The ContactPoint object to format.
 * @returns - The formatted contact information.
 */
function formatContact(contacts: ContactPoint[]): string {
  return contacts
    .map((contact) => {
      if (contact.system === "phone") {
        return `${contact.use}: ${contact.value}`;
      } else if (contact.system === "email") {
        return contact.value;
      }
    })
    .join("\n");
}

/**
 * Formats the identifiers of a FHIR Identifier object.
 * @param identifier - The Identifier object to format.
 * @returns The formatted identifiers.
 */
function formatIdentifier(identifier: Identifier[]): string {
  return identifier
    .map((id) => {
      let idType = id.type?.coding?.[0].display ?? "";
      if (idType === "") {
        idType = id.type?.text ?? "";
      }

      return `${idType}: ${id.value}`;
    })
    .join("\n");
}

/**
 * Calculates the age of a patient based on their birth date.
 * @param patient - The patient to calculate the age for.
 * @returns The age of the patient.
 */
export function calculatePatientAge(patient: Patient): number | undefined {
  if (patient.birthDate) {
    const patientDOB = new Date(patient.birthDate);
    const today = new Date();
    return dateFns.differenceInYears(today, patientDOB);
  }
}
