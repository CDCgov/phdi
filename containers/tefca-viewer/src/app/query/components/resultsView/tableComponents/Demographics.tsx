import { Patient } from "fhir/r4";
import { DataDisplayInfo } from "@/app/utils";
import { DataDisplay } from "@/app/utils";
import * as dateFns from "date-fns";
import { evaluate } from "fhirpath";

import {
  formatName,
  formatAddress,
  formatContact,
  formatIdentifier,
  formatDate,
} from "../../../../format-service";

/**
 * Displays the demographic information of a patient.
 * @param {Patient} patient - The patient to display demographic information for.
 * @returns {React.FC} The Demographics component.
 */
export interface DemographicsProps {
  patient: Patient;
}

/**
 * Displays the demographic information of a patient.
 * @param props - The props for the Demographics component.
 * @param props.patient - The patient resource to display demographic information for.
 * @returns The Demographics component.
 */
const Demographics: React.FC<DemographicsProps> = ({ patient }) => {
  const demographicData = formatDemographics(patient);

  return (
    <div>
      {demographicData.map((item) => (
        <DataDisplay item={item} key={item.title} />
      ))}
    </div>
  );
};

export default Demographics;

/**
 * Formats the demographic information of a patient.
 * @param patient - The patient to format demographic information for.
 * @returns The formatted demographic information as an array of DisplayData objects.
 */
function formatDemographics(patient: Patient): DataDisplayInfo[] {
  const demographicData: DataDisplayInfo[] = [
    {
      title: "Patient Name",
      value: formatName(patient.name ?? []),
    },
    {
      title: "DOB",
      value: formatDate(patient.birthDate),
    },
    {
      title: "Current Age",
      value: calculatePatientAge(patient)?.toString(),
    },
    {
      title: "Sex",
      value: patient.gender ?? "",
    },
    {
      title: "Race",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Ethnicity",
      value: evaluate(
        patient,
        "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.first().valueCoding.display",
      )[0],
    },
    {
      title: "Tribal Affiliation",
      value: evaluate(
        patient,
        "Patient.extension.where(url='http: //hl7.org/fhir/us/ecr/StructureDefinition/us-ph-tribal-affiliation-extension').extension.where(url='TribeName').value.display",
      )[0],
    },
    {
      title: "Preferred Language",
      value: evaluate(
        patient,
        "Patient.communication.first().language.coding.first().display",
      )[0],
    },
    {
      title: "Address",
      value: formatAddress(patient.address ?? []),
    },
    {
      title: "Contact",
      value: formatContact(patient.telecom ?? []),
    },
    {
      title: "Patient Identifiers",
      value: formatIdentifier(patient.identifier ?? []),
    },
  ];

  return demographicData;
}

/**
 * Calculates the age of a patient based on their birth date.
 * @param patient - The patient to calculate the age for.
 * @returns The age of the patient.
 */
export function calculatePatientAge(patient: Patient): number | undefined {
  if (patient.birthDate) {
    const patientDOB = new Date(patient.birthDate);
    const today = new Date();
    return dateFns.differenceInYears(today, patientDOB);
  }
}
