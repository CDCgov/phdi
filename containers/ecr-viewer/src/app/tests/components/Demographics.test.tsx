import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import Demographics from "../../view-data/components/Demographics";

describe("Demographics", () => {
  let container: HTMLElement;
  beforeAll(() => {
    const demographicsData = [
      {
        title: "Patient Name",
        value: "Test patient",
      },
      { title: "DOB", value: "06/01/1996" },
      { title: "Current Age", value: "27" },
      { title: "Vital Status", value: "Alive" },
      { title: "Sex", value: "female" },
      { title: "Race", value: "Asian/Pacific Islander" },
      {
        title: "Ethnicity",
        value: "Not Hispanic or Latino",
      },
      {
        title: "Tribal",
        value: "test",
      },
      {
        title: "Preferred Language",
        value: "test",
      },
      {
        title: "Patient Address",
        value: "test address",
      },
      {
        title: "County",
        value: "test",
      },
      { title: "Contact", value: "test contact" },
      {
        title: "Emergency Contact",
        value: "N/A",
      },
      {
        title: "Patient IDs",
        value: "123-4567-890",
      },
    ];
    container = render(
      <Demographics demographicsData={demographicsData} />,
    ).container;
  });
  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });
  it("should pass accessibility test", async () => {
    expect(await axe(container)).toHaveNoViolations();
  });
});
