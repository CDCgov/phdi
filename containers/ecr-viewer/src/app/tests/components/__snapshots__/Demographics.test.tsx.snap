// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demographics should match snapshot 1`] = `
<div>
  <div
    class="margin-top-0"
  >
    <div
      class="padding-bottom-3"
    >
      <div
        class="usa-summary-box__body"
      >
        <h4
          class="usa-summary-box__heading padding-y-105"
          id="demographics"
        >
          Demographics
        </h4>
        <div
          class="usa-summary-box__text"
        >
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Patient Name
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Test patient
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                DOB
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                06/01/1996
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Current Age
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                27
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Vital Status
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Alive
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Sex
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                female
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Race
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Asian/Pacific Islander
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Ethnicity
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Not Hispanic or Latino
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Tribal
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                test
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Preferred Language
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                test
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Patient Address
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                test address
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                County
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                test
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Contact
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                test contact
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Emergency Contact
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                N/A
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Patient IDs
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                123-4567-890
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
